Enable Trino to detect and fail queries that have tasks stuck in long running JONI parsing.

leetcode-1533 opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->
Enable Trino to detect and fail tasks that are stuck in long running JONI parsing. 

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->


> Is this change a fix, improvement, new feature, refactoring, or other?

Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

 Query Engine

> How would you describe this change to a non-technical end user or system administrator?

Trino is a time shared multi-tenant system. For context switching, trino relies on the threadpool workers cooperatively yield itself to the scheduling logic. It is not like an operating system which used a hard interrupt signal to force the process to do the context switch.

Most of the split processing can be finished in a relative short interval, whereas the JONI processing is an exception.

Furthermore, some of the trino features: gathering statistics, relaying on the trino to do callback function(to execute splitFinished() after yield) after the context switch, since there is no context switch in this case, trino can't gather accurate CPU usage for that split. 

This PR allows trino to detect and fail tasks that are stuck in long running JONI parsing. 

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

Enable Trino to detect and fail queries that have tasks stuck in long running JONI parsing.

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:
```diff
@@ -463,4 +464,18 @@ public TaskManagerConfig setTaskYieldThreads(int taskYieldThreads)
+    }
+
+    @Config("task.interrupt-runaway-splits-timeout")
+    @ConfigDescription("Interrupt runaway split threads after this timeout if the task is stuck in certain allow listed places")
```
```suggestion
    @ConfigDescription("Interrupt task processing thread after this timeout if the thread is stuck in certain external libraries used by Trino functions")
```

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -82,7 +87,10 @@
     // print out split call stack if it has been running for a certain amount of time
     private static final Duration LONG_SPLIT_WARNING_THRESHOLD = new Duration(600, TimeUnit.SECONDS);
-
+    private static final Predicate<List<StackTraceElement>> DEFAULT_INTERRUPTIBLE_SPLIT_PREDICATE =
```
I would drop the `DEFAULT` and rename to `POTENTIALLY_STUCK_SPLIT_PREDICATE`

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -82,7 +87,10 @@
+    private static final Predicate<List<StackTraceElement>> DEFAULT_INTERRUPTIBLE_SPLIT_PREDICATE =
+            elements -> elements.stream().anyMatch(element -> element.getClassName().equals(JoniRegexpFunctions.class.getName()));
+    // checking frequency for long-running splits
+    private static final Duration DEFAULT_INTERRUPT_SPLIT_INTERVAL = new Duration(60, TimeUnit.SECONDS);
```
STUCK_SPLIT_DETECTION_FREQUENCY

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -187,20 +204,29 @@ public TaskExecutor(
             int guaranteedNumberOfDriversPerTask,
             int maximumNumberOfDriversPerTask,
             VersionEmbedder versionEmbedder,
+            Duration interruptRunawaySplitsTimeout,
```
interrupStuckSplitsTimeout/interruptStuckSplitsDelay (should match config option)

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -187,20 +204,29 @@ public TaskExecutor(
             int maximumNumberOfDriversPerTask,
             VersionEmbedder versionEmbedder,
+            Duration interruptRunawaySplitsTimeout,
+            Predicate<List<StackTraceElement>> interruptibleSplitPredicate,
```
`potentiallyStuckSplitPredicate`

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -82,7 +87,10 @@
-
+    private static final Predicate<List<StackTraceElement>> DEFAULT_INTERRUPTIBLE_SPLIT_PREDICATE =
+            elements -> elements.stream().anyMatch(element -> element.getClassName().equals(JoniRegexpFunctions.class.getName()));
+    // checking frequency for long-running splits
```
drop comment

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -187,20 +204,29 @@ public TaskExecutor(
+
         this.versionEmbedder = requireNonNull(versionEmbedder, "versionEmbedder is null");
 
+        this.interruptRunawaySplitsTimeout = interruptRunawaySplitsTimeout;
+        this.interruptibleSplitPredicate = interruptibleSplitPredicate;
+        this.interruptSplitInterval = interruptSplitInterval;
```
Rename fields to match args. Make all fields not-null. Use `Optional` when applicable.

phd3 replied:
I'd just put all this in one object which can be made optional

losipiuk commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -217,6 +243,10 @@ public synchronized void start()
             addRunnerThread();
         }
+        if (interruptRunawaySplitsTimeout != null) {
+            long interval = (long) interruptSplitInterval.getValue(TimeUnit.SECONDS);
```
intervalInSeconds

leetcode-1533 replied:
done,

findepi commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,6 +487,39 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+
+                if (interruptibleSplitPredicate.test(stackTraceElements)) {
+                    log.warn("Interrupting runaway split matching the predicate " + splitInfo.getSplitInfo());
+                    splitInfo.getThread().interrupt();
```
is this racy?

losipiuk replied:
That is a good point. 
We need coordination between a code that issues an interruption and runner threads. So we know that we are only interrupting the runner thread if it is still executing the query we observed in the stack trace (and not after the thread picked up split from another query).

Also does interrupting thread which executes within Joni library actually makes the call from the thread return?

phd3 replied:
> Also does interrupting thread which executes within Joni library actually makes the call from the thread return?

currently seems that it catches the interruptedexception [silently](https://github.com/airlift/joni/blob/master/src/org/joni/Matcher.java#L96) with an error value that does not seem handled. would need something like https://github.com/prestodb/presto/pull/16109 . but this may not be guaranteed based on this comment https://github.com/prestodb/presto/pull/16109#pullrequestreview-661488484

phd3 commented on core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:
```diff
@@ -69,6 +69,7 @@
 
     private Duration statusRefreshMaxWait = new Duration(1, TimeUnit.SECONDS);
     private Duration infoUpdateInterval = new Duration(3, TimeUnit.SECONDS);
+    private Duration interruptRunawaySplitsTimeout = new Duration(600, TimeUnit.SECONDS);
```
We also have a warning about the runaway splits, currently hardcoded. IMO we should validate the two thresholds are coherent (i.e. warning <= timeout), which would require  making the warning configurable too

leetcode-1533 replied:
got it, will configure: "LONG_SPLIT_WARNING_THRESHOLD"

leetcode-1533 replied:
agree we should make  warning <= timeout, because if the thread runtime for the split > than timeout, the system will interrupt the split.

leetcode-1533 replied:
I also felt "/v1/maxActiveSplits" is unnecessary.. Since this PR will print out the run away splits in the log..

phd3 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -164,20 +176,25 @@ public TaskExecutor(TaskManagerConfig config, VersionEmbedder versionEmbedder, M
     public TaskExecutor(int runnerThreads, int minDrivers, int guaranteedNumberOfDriversPerTask, int maximumNumberOfDriversPerTask, Ticker ticker)
     {
-        this(runnerThreads, minDrivers, guaranteedNumberOfDriversPerTask, maximumNumberOfDriversPerTask, testingVersionEmbedder(), new MultilevelSplitQueue(2), ticker);
+        this(runnerThreads, minDrivers, guaranteedNumberOfDriversPerTask, maximumNumberOfDriversPerTask, testingVersionEmbedder(),
```
nit: put everything on new line for consistency with other places

leetcode-1533 replied:
done

phd3 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -164,20 +176,25 @@ public TaskExecutor(TaskManagerConfig config, VersionEmbedder versionEmbedder, M
     public TaskExecutor(int runnerThreads, int minDrivers, int guaranteedNumberOfDriversPerTask, int maximumNumberOfDriversPerTask, MultilevelSplitQueue splitQueue, Ticker ticker)
     {
-        this(runnerThreads, minDrivers, guaranteedNumberOfDriversPerTask, maximumNumberOfDriversPerTask, testingVersionEmbedder(), splitQueue, ticker);
+        this(runnerThreads, minDrivers, guaranteedNumberOfDriversPerTask, maximumNumberOfDriversPerTask, testingVersionEmbedder(),
```
same formatting comment

phd3 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -187,20 +204,29 @@ public TaskExecutor(
         checkArgument(guaranteedNumberOfDriversPerTask > 0, "guaranteedNumberOfDriversPerTask must be at least 1");
         checkArgument(maximumNumberOfDriversPerTask > 0, "maximumNumberOfDriversPerTask must be at least 1");
         checkArgument(guaranteedNumberOfDriversPerTask <= maximumNumberOfDriversPerTask, "guaranteedNumberOfDriversPerTask cannot be greater than maximumNumberOfDriversPerTask");
+        checkArgument(interruptRunawaySplitsTimeout.getValue(TimeUnit.SECONDS) >= 1.0, "interruptRunawaySplitsTimeout must be at least 1 second");
```
is `1s` based off of SPLIT_RUN_QUANTA? if so let's just use that

leetcode-1533 replied:
ok.

phd3 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,6 +487,39 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+
+                    String splitThreadStack = toStackString(stackTraceElements);
+                    String splitThreadId = splitInfo.getThreadId();
+                    log.warn(String.format("Split %s has been long running %n%s", splitInfo.getSplitInfo(), splitThreadStack));
```
nit: static import `format`

phd3 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,6 +487,39 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+                    splitInfo.setPrinted();
+
+                    String splitThreadStack = toStackString(stackTraceElements);
+                    String splitThreadId = splitInfo.getThreadId();
```
unused?

phd3 commented on core/trino-main/src/test/java/io/trino/execution/executor/TestTaskExecutor.java:
```diff
@@ -491,6 +493,39 @@ public void testMinDriversPerTaskWhenTargetConcurrencyIncreases()
+        try {
+            TaskId taskId = new TaskId(new StageId("foo", 0), 0, 0);
+            TaskHandle taskHandle = taskExecutor.addTask(
+                    taskId, () -> 1.0,
```
nit: parameter on new line

leetcode-1533 replied:
done

leetcode-1533 commented:
Hey, I found out the issue is very similar to https://github.com/trinodb/trino/pull/7213, in which we want to limit the total time for SQLQueryExecution. In that case, the thread is reused between queries, in our case, the thread is reused between tasks. In either case, I abstract the following programming pattern:



1. When assigning task to the thread:

`AtomicReference<Thread> sharedThread = new AtomicReference<>(currentThread())
`

2. After 1, when processing:

```
try {
	process using the shared thread for the assigned task
} finally {
	synchronized (exectuorObject) {
		sharedThread = null;
		Thread.interrupt();
	}
}

```
3. When interrupting:

```
{
	if can interrupt();
	synchronized (exectuorObject) {
// double check once get into the critical section, 
// to avoid race condition happening at the short interval between the check and get into the race condition. 
                if can interrupt();  
		thread = sharedThread.get();
		if (thread != null) {
			thread.interrupt();
		}
	}
}
```
 

From my personally understanding, the AtomicReference can also be replaced by volatile keyword.

findepi commented:
@leetcode-1533 would you mind checking the build status?

findepi commented:
can you please make sure you filled the CLA? For more information, see https://github.com/trinodb/cla.

electrum commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+
+                    String splitThreadStack = toStackString(stackTraceElements);
+                    String splitThreadId = splitInfo.getThreadId();
+                    log.warn(String.format("Split %s has been long running %n%s", splitInfo.getSplitInfo(), splitThreadStack));
```
The logger methods already take a format string, so no need for `format()`

electrum commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+                }
+
+                if (interruptibleSplitPredicate.test(stackTraceElements)) {
+                    log.warn("Potentially interrupting runaway split matching the predicate " + splitInfo.getSplitInfo());
```
Use the logger format string `%s` rather than concatenation

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
             implements Runnable
     {
         private final long runnerId = NEXT_RUNNER_ID.getAndIncrement();
+        public AtomicReference<RunningSplitInfo> currentThreadSplit = new AtomicReference<>();
```
private final

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+                if (interruptibleSplitPredicate.test(stackTraceElements)) {
+                    log.warn("Potentially interrupting runaway split matching the predicate " + splitInfo.getSplitInfo());
+
+                    synchronized (splitInfo.getTaskRunner()) {
```
Instead of synchronizing on the task runner here I would recommend having a `splitInfo.getTaskRunner().interruptSplit(splitInfo)` to have the synchronization well encapsulated

leetcode-1533 replied:
done.

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+                    splitInfo.setPrinted();
+
+                    String splitThreadStack = toStackString(stackTraceElements);
+                    String splitThreadId = splitInfo.getThreadId();
```
Seems to be unused

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+            if (duration.compareTo(interruptRunawaySplitsTimeout) > 0) {
+                List<StackTraceElement> stackTraceElements = asList(splitInfo.getThread().getStackTrace());
+
+                if (!splitInfo.printed) {
```
`isPrinted()`

leetcode-1533 replied:
done.

jhlodin commented on docs/src/main/sphinx/admin/properties-task.rst:
```diff
@@ -124,3 +124,14 @@ of additional CPU for parallel writes. Some connectors can be bottlenecked on CP
+``task.interrupt-runaway-splits-timeout``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* **Type:** ``duration``
```
should be :ref:`prop-type-duration` matching other properties

leetcode-1533 replied:
done.

jhlodin commented on docs/src/main/sphinx/admin/properties-task.rst:
```diff
@@ -124,3 +124,14 @@ of additional CPU for parallel writes. Some connectors can be bottlenecked on CP
+* **Minimum value:** ``1s``
+* **Default value:** ``10m``
+
+Timeout for interrupting split threads blocked without yielding control.
```
Maybe "The length of time Trino waits for a blocked split processing thread before interrupting the thread." if that is still technically accurate.

jhlodin commented on docs/src/main/sphinx/admin/properties-task.rst:
```diff
@@ -124,3 +124,14 @@ of additional CPU for parallel writes. Some connectors can be bottlenecked on CP
+* **Default value:** ``10m``
+
+Timeout for interrupting split threads blocked without yielding control.
+Only threads blocked in specific locations are interrupted. Currently this is just threads
+blocked in the Joni regular expression library.
```
Shoudn't use "currently" in the docs - even if we plan to change it later we can't predict the future in documentation until it happens.

I'd suggest replacing both sentences with "Only applies to threads that are blocked by the third-party Joni regular expression library"

leetcode-1533 replied:
done.

jhlodin reviewed (commented):
Some comments on the documentation

leetcode-1533 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,10 +488,50 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+                        if (synchronizedRunningSplit == splitInfo) {
+                            splitInfo.getThread().interrupt();
+                        }
+                    }
```
fail the task.

tangjiangling commented:
Hey guys, what is the current state of this PR? (we have a similar problem in our production environment)

tangjiangling commented:
I have a question: now that we proactively let the task fail, what changes are required for the user to make the query run successfully? 
(e.g., can we switch the regular library via session property, which is not supported yet).

arhimondr commented:
@tangjiangling It has comments that haven't been addressed. We should be able to merge once the comments are resolved.

leetcode-1533 commented:
Hi, thanks for asking! Yes, should be update both PR by the end of today!

tangjiangling commented:
> Yes, should be update both PR by the end of today!

Superseded by #13042, sorry, I didn't see your reply, you can update another PR today.

leetcode-1533 commented on core/trino-main/src/test/java/io/trino/execution/executor/TestTaskExecutor.java:
```diff
@@ -517,6 +552,44 @@ private static void waitUntilSplitsStart(List<TestingJob> splits)
+        {
+            while (true) {
+                try {
+                    Thread.sleep(1);
```
done

leetcode-1533 commented:
Hi, thanks for helping! I have updated the PR.

I deeply apologize for the delay due to personal errands!! From now on, I will be more responsive to PR feedback.

Significant changes:
1. Per David comment removed atomic wrapping for currentThreadSplitInfo, since it is only used in critical section.
2. Per Arosa comment, refactored interrupt related functions and added TaskRunner.interruptIfMatch(RunningSplitInfo) and RunningSplitInfo.interruptThreadIfRunning(), which I think is less confusing than "splitInfo.getTaskRunner().interruptSplit(splitInfo)".

I felt it is more straight forward since it passed in argument as "this" rather than a splitInfo member in the for loop.

3. Per tangjiangling suggestions, added class StuckSplitInterrupter.

4. Fixed documents.


Other things to mention:
1. I intentionally didn't change the API "/v1/maxActiveSplits" for backward compatibility. Though I think given we added the interrupter, this API is kind of duplicate. Trino admin can check the log rather than using the API to find out the run away splits. 

Currently, the interrupter will print out stack trace for all long running splits but only actually interrupt thread that are using Joni.

Please let me know if we want to remove the API.

tangjiangling commented:
> I deeply apologize for the delay due to personal errands!! From now on, I will be more responsive to PR feedback.

Never mind.

> Thank you for your pull request and welcome to our community. We could not parse the GitHub identity of the following contributors: yluan.

You need to update `user.name` and `user.email` to match your previously signed [CLA](https://github.com/trinodb/cla/blob/master/contributors#L385) account (I see you have signed the CLA), otherwise it will not be possible to merge PRs.

tangjiangling commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,11 +484,85 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
         return null;
     }
 
+    @VisibleForTesting
```
Remove as we only use it in the test code (Per the `VisibleForTesting` doc: `Annotates a program element that exists, or is more widely visible than otherwise necessary, only for use in test code`).

leetcode-1533 replied:
Hey, this is needed in unit test: testTaskExecutorInterruptRunawaySplit().

I can't make this method private, since currently the static inner class "StuckSplitInterrupter" is private and not accessible for the unit test(which is intentional, since StuckSplitInterrupter is an internal logic of taskExecutor).

tangjiangling commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -66,22 +70,32 @@
 
     // print out split call stack if it has been running for a certain amount of time
-    private static final Duration LONG_SPLIT_WARNING_THRESHOLD = new Duration(600, TimeUnit.SECONDS);
+    private static final Duration LONG_SPLIT_WARNING_THRESHOLD = new Duration(600, SECONDS);
```
We also need to make `LONG_SPLIT_WARNING_THRESHOLD` configurable and need to verify that `LONG_SPLIT_WARNING_THRESHOLD <= interruptStuckSplitsTimeout` (you can refer to #13042)

leetcode-1533 replied:
@arhimondr How do you think, should we keep API /v1/maxActiveSplits and make the threshold configurable via config?

tangjiangling reviewed (commented):
skimmed

tangjiangling commented:
> I intentionally didn't change the API "/v1/maxActiveSplits" for backward compatibility.

Agree, although the naming of the interface confuses me (maybe `/v1/longRunningSplits` or `/v1/stuckSplits` would be better for me)

> Though I think given we added the interrupter, this API is kind of duplicate. Trino admin can check the log rather than using the API to find out the run away splits.

I think we still need this because this feature gives us another, simpler way to find long-running splits (we are now relying on this interface to see what logic those long-running splits are actually processing)

leetcode-1533 commented:
> > I deeply apologize for the delay due to personal errands!! From now on, I will be more responsive to PR feedback.
> 
> Never mind.
> 
> > Thank you for your pull request and welcome to our community. We could not parse the GitHub identity of the following contributors: yluan.
> 
> You need to update `user.name` and `user.email` to match your previously signed [CLA](https://github.com/trinodb/cla/blob/master/contributors#L385) account (I see you have signed the CLA), otherwise it will not be possible to merge PRs.

Hey, thanks: it has been sighed.

leetcode-1533 commented:
> > I intentionally didn't change the API "/v1/maxActiveSplits" for backward compatibility.
> 
> Agree, although the naming of the interface confuses me (maybe `/v1/longRunningSplits` or `/v1/stuckSplits` would be better for me)
> 
> > Though I think given we added the interrupter, this API is kind of duplicate. Trino admin can check the log rather than using the API to find out the run away splits.
> 
> I think we still need this because this feature gives us another, simpler way to find long-running splits (we are now relying on this interface to see what logic those long-running splits are actually processing)

Thanks, since it is an API change, let's prefer not to alter the naming for maximum backward compatibility.

I will follow https://github.com/trinodb/trino/pull/13042 and make the warning thresholds configurable. And I think we shouldn't add this config to the interrupter: it is not used in the interrupter, but actually being directly used TaskExecutor's public method: getMaxActiveSplitsInfo().

Still, the information are duplicate with the logging message
from interruptStuckSplits(), it printed stackTrace for all splits (including non-Joni related threads) as well:
  log.warn("Split %s has been long running %n%s", splitInfo.getSplitInfo(), toStackString(stackTraceElements));

phd3 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,11 +484,85 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+            return stackString;
+        }
+
+        private void interruptStuckSplits(SortedSet<RunningSplitInfo> runningSplitInfos, Ticker ticker)
```
is this comment addressed? https://github.com/trinodb/trino/pull/12479#discussion_r881715903

tangjiangling replied:
I believe not yet.

arhimondr reviewed (commented):
I'm trying to understand in what scenario this PR is useful. 

I don't think we should try to kill a query as long as the end user is willing to wait for it. When somebody writes a query that has a very long running regular expression in it it feels like we should let it run and shouldn't try to impose any driver level restrictions on the run time. If a limit on run time is necessary it can be configured with `query.max-run-time`, `query.max-execution-time`, `query.max-cpu-time`, etc.

When query is cancelled by the end user (or fails on timeout) the coordinator will cancel all running tasks. Upon task cancellation a worker interrupts all running splits for a given task. If there's evidence it doesn't happen in certain circumstances - it's a bug and it must be fixed.

Sometimes, when the code that is run by the driver is not interruptible (addressed for regular expressions in https://github.com/trinodb/trino/pull/13064) it is possible that a worker may leave some splits behind. However if the interrupt on task cancellation wasn't able to interrupt why do we think a scheduled interrupt is going to help? 

CC: @trinodb/maintainers

losipiuk commented:
> I'm trying to understand in what scenario this PR is useful.
> 
> I don't think we should try to kill a query as long as the end user is willing to wait for it. When somebody writes a query that has a very long running regular expression in it it feels like we should let it run and shouldn't try to impose any driver level restrictions on the run time. If a limit on the run time is necessary it can be configured with `query.max-run-time`/`query.max-execution-time`/`query.max-cpu-time`, etc.
> 
> When query is cancelled by the end user (or fails on timeout) the coordinator will cancel all running tasks. Upon task cancellation a worker interrupts all running splits for a given task. Is there evidence it doesn't happen in certain circumstances? (If that's the case than it is a bug and we must fix it)
> 
> Sometimes, when code that is ran by the driver is not interruptible (addressed for regular expressions in #13064) it is possible that a worker may leave some splits behind. However if the interrupt on task cancellation doesn't interrupt why do we think a scheduled interrupt will?
> 
> CC: @trinodb/maintainers

Makes sense.
It feels the only reason when it is useful is when generally constraining query runtime is not an option. And we do not want the broken queries to consume the cluster. 
It may not even be about CPU consumption (I am not sure if JONI gets stuck with high CPU usage) but it may use query slot, and prevent other queries from starting up.

leetcode-1533 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,11 +484,85 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
+                    .map(Object::toString)
+                    .collect(joining(lineSeparator()));
+
+            return stackString;
```
done.

leetcode-1533 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -478,10 +579,14 @@ public void run()
-                        runningSplitInfos.add(splitInfo);
+                        RunningSplitInfo splitInfo = new RunningSplitInfo(ticker.read(), threadId, Thread.currentThread(), split, this);
+                        synchronized (this) {
+                            currentThreadSplitInfo = splitInfo;
```
Thanks, that is a good point!

Given this, and https://github.com/trinodb/trino/pull/12479#discussion_r881715903, I am thinking using the taskManager.abortTask(taskId) instead of implementing its own interrupt mechanism.

This can avoid the race condition that the interrupt thread and other threads concurrently trying to abort a task.

How do you think?

arhimondr replied:
Yeah, I think it's a good idea. You can try to use either `SqlTask#failed` or `SqlTaskManager#failTask`. These methods will allow you to provide an exception to be used to fail task with where you can include a meaningful error message.

phd3 commented:
Good point @arhimondr

>  However if the interrupt on task cancellation doesn't interrupt why do we think a scheduled interrupt will?

Note that before #13064, task cancellation interrupt was ineffective in the JONI case, and that's how we observed the issue. After that one, if the user cancels, the splits should get killed. However, this is only relevant for "failed" queries.  

> I don't think we should try to kill a query as long as the end user is willing to wait for it. When somebody writes a query that has a very long running regular expression in it it feels like we should let it run and shouldn't try to impose any driver level restrictions on the run time. If a limit on the run time is necessary it can be configured with query.max-run-time/query.max-execution-time/query.max-cpu-time, etc.

Even though constraining `query.max-run-time` is an option, such a limit would be much higher practically (potentially in  10s of minutes or hours). And during that time, such "bad" queries can quickly and significantly reduce the effective concurrency of the cluster. This'd not be desirable.

`query.max-cpu-time` can be useful here, however, the cpu recording for an in-progress split currently happens at the end, so it's not really accounted for in the joni case. https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/Driver.java#L341 And that's primarily where the "driver level restriction" is coming from. 
So another option would be to somehow track such "in-progress CPU", not sure how complex that'd be though.

arhimondr commented:
@phd3 Actually yeah, you are right. For the entire run time the entire cluster will be effectively blocked by a "bad" query

arhimondr commented:
@leetcode-1533 @phd3 Yeah, this is a very good point. For the entire duration of a "bad" query the entire cluster will be blocked. I'm being slow. We certainly need it.

leetcode-1533 commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -457,11 +484,85 @@ private synchronized PrioritizedSplitRunner pollNextSplitWorker()
 
+    @VisibleForTesting
+    public static Optional<StuckSplitInterrupter> createStuckSplitInterrupter(
+            boolean enableInterruptStuckSplits,
```
ok, thanks!

leetcode-1533 commented:
Hi,

Refactored the code based on failing the task instead of directly interrupting the runner thread.


Sample failed query(from CLI):
```
trino> SELECT regexp_replace(veryLongStr, '\\((.*,)+(.*\\))', '\\($2\\)') as requesttarget FROM memory.memoryTables.testTable;
    ->

Query 20220713_020122_00003_mffv4, FAILED, 1 node
Splits: 26 total, 9 done (34.62%)
17.07 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20220713_020122_00003_mffv4 failed: Task 20220713_020122_00003_mffv4.1.0.0 containing long running stuck split get cancelled by the system
```

Sample log:
```
022-07-12T19:01:39.236-0700	WARN	task-management-4	io.trino.execution.SqlTaskManager	Split Split 20220713_020122_00003_mffv4.1.0.0-0 MemorySplit{tableHandle=0, partNumber=0, totalPartsPerWorker=10} (start = 1.26293092989791E8, wall = 16858 ms, cpu = 0 ms, wait = 0 ms, calls = 1) is long running of stackTrace: app//io.airlift.jcodings.specific.NonStrictUTF8Encoding.length(NonStrictUTF8Encoding.java:30)
app//io.airlift.joni.ByteCodeMachine.opAnyCharStarPeekNext(ByteCodeMachine.java:925)
app//io.airlift.joni.ByteCodeMachine.matchAt(ByteCodeMachine.java:232)
app//io.airlift.joni.Matcher.matchCheck(Matcher.java:304)
app//io.airlift.joni.Matcher.searchInterruptible(Matcher.java:457)
app//io.trino.operator.scalar.JoniRegexpFunctions.getSearchingOffset(JoniRegexpFunctions.java:416)
app//io.trino.operator.scalar.JoniRegexpFunctions.regexpReplace(JoniRegexpFunctions.java:114)
java.base@11.0.15/java.lang.invoke.LambdaForm$DMH/0x00000008007a6040.invokeStaticInit(LambdaForm$DMH)
java.base@11.0.15/java.lang.invoke.LambdaForm$MH/0x0000000802d78840.invoke(LambdaForm$MH)
java.base@11.0.15/java.lang.invoke.Invokers$Holder.linkToTargetMethod(Invokers$Holder)
io.trino.$gen.PageProjectionWork_20220713_020122_38.evaluate(Unknown Source)
io.trino.$gen.PageProjectionWork_20220713_020122_38.process(Unknown Source)
app//io.trino.operator.project.DictionaryAwarePageProjection$DictionaryAwarePageProjectionWork.processInternal(DictionaryAwarePageProjection.java:185)
app//io.trino.operator.project.DictionaryAwarePageProjection$DictionaryAwarePageProjectionWork.process(DictionaryAwarePageProjection.java:121)
app//io.trino.operator.project.PageProcessor$ProjectSelectedPositions.processBatch(PageProcessor.java:336)
app//io.trino.operator.project.PageProcessor$ProjectSelectedPositions.process(PageProcessor.java:208)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils.lambda$flatten$7(WorkProcessorUtils.java:296)
app//io.trino.operator.WorkProcessorUtils$$Lambda$10552/0x0000000802d7e040.process(Unknown Source)
app//io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils.lambda$flatten$7(WorkProcessorUtils.java:296)
app//io.trino.operator.WorkProcessorUtils$$Lambda$10552/0x0000000802d7e040.process(Unknown Source)
app//io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:338)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:325)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
app//io.trino.operator.WorkProcessorUtils.lambda$processStateMonitor$3(WorkProcessorUtils.java:219)
app//io.trino.operator.WorkProcessorUtils$$Lambda$10556/0x0000000802d7d040.process(Unknown Source)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:240)
app//io.trino.operator.WorkProcessorUtils.lambda$finishWhen$4(WorkProcessorUtils.java:234)
app//io.trino.operator.WorkProcessorUtils$$Lambda$10558/0x0000000802d7c840.process(Unknown Source)
app//io.trino.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:391)
app//io.trino.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:150)
app//io.trino.operator.Driver.processInternal(Driver.java:410)
app//io.trino.operator.Driver.lambda$process$10(Driver.java:313)
app//io.trino.operator.Driver$$Lambda$10121/0x0000000802ba9440.get(Unknown Source)
app//io.trino.operator.Driver.tryWithLock(Driver.java:698)
app//io.trino.operator.Driver.process(Driver.java:305)
app//io.trino.operator.Driver.processForDuration(Driver.java:276)
app//io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:740)
app//io.trino.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:164)
app//io.trino.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:490)
io.trino.$gen.Trino_dev____20220713_020007_2.run(Unknown Source)
java.base@11.0.15/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
java.base@11.0.15/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
java.base@11.0.15/java.lang.Thread.run(Thread.java:829)
2022-07-12T19:01:39.243-0700	ERROR	remote-task-callback-2	io.trino.execution.scheduler.PipelinedStageExecution	Pipelined stage execution for stage 20220713_020122_00003_mffv4.1 failed
io.trino.spi.TrinoException: Task 20220713_020122_00003_mffv4.1.0.0 containing long running stuck split get cancelled by the system
	at io.trino.execution.SqlTaskManager$StuckSplitTaskInterrupter.failStuckSplitTasks(SqlTaskManager.java:732)
	at io.trino.execution.SqlTaskManager.lambda$failStuckSplitTasks$13(SqlTaskManager.java:664)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.trino.execution.SqlTaskManager.failStuckSplitTasks(SqlTaskManager.java:664)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)


2022-07-12T19:01:39.245-0700	ERROR	stage-scheduler	io.trino.execution.scheduler.SqlQueryScheduler	Failure in distributed stage for query 20220713_020122_00003_mffv4
io.trino.spi.TrinoException: Task 20220713_020122_00003_mffv4.1.0.0 containing long running stuck split get cancelled by the system
	at io.trino.execution.SqlTaskManager$StuckSplitTaskInterrupter.failStuckSplitTasks(SqlTaskManager.java:732)
	at io.trino.execution.SqlTaskManager.lambda$failStuckSplitTasks$13(SqlTaskManager.java:664)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.trino.execution.SqlTaskManager.failStuckSplitTasks(SqlTaskManager.java:664)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)


2022-07-12T19:01:39.247-0700	ERROR	stage-scheduler	io.trino.execution.StageStateMachine	Stage 20220713_020122_00003_mffv4.1 failed
io.trino.spi.TrinoException: Task 20220713_020122_00003_mffv4.1.0.0 containing long running stuck split get cancelled by the system
	at io.trino.execution.SqlTaskManager$StuckSplitTaskInterrupter.failStuckSplitTasks(SqlTaskManager.java:732)
	at io.trino.execution.SqlTaskManager.lambda$failStuckSplitTasks$13(SqlTaskManager.java:664)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.trino.execution.SqlTaskManager.failStuckSplitTasks(SqlTaskManager.java:664)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)


2022-07-12T19:01:39.266-0700	INFO	dispatcher-query-0	io.trino.event.QueryMonitor	TIMELINE: Query 20220713_020122_00003_mffv4 :: Transaction:[058ef3de-b0fd-4e9d-b4e7-28cf02bdc351] :: elapsed 17068ms :: planning 95ms :: waiting 37ms :: scheduling 118ms :: running 16846ms :: finishing 9ms :: begin 2022-07-12T19:01:22.180-07:00 :: end 2022-07-12T19:01:39.248-07:00

```

jhlodin commented on docs/src/main/sphinx/admin/properties-task.rst:
```diff
@@ -124,3 +124,42 @@ of additional CPU for parallel writes. Some connectors can be bottlenecked on CP
+* **Minimum value:** ``1s``
+* **Default value:** ``10m``
+
+Print out call stacks at ``/v1/maxActiveSplits`` endpoint and generate JMX metrics for splits running longer than the threshold.
```
Please word wrap at 80 characters, here and elsewhere in the file

jhlodin reviewed (approved):
Docs LGTM, just need to word wrap at 80 characters

phd3 commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -211,7 +259,12 @@ public void start()
+
+        stuckSplitTaskInterrupter.ifPresent(interrupter -> {
+            long intervalSeconds = interrupter.getStuckSplitsDetectionInterval().roundTo(SECONDS);
+            taskManagementExecutor.scheduleAtFixedRate(this::failStuckSplitTasks, 0, intervalSeconds, SECONDS);
```
let's put try-catch with a generic throwable like the above, because schedule* methods supress future executions on any sort of exception throwing

phd3 commented on core/trino-main/src/test/java/io/trino/execution/TestSqlTaskManager.java:
```diff
@@ -374,4 +435,43 @@ public URI createMemoryInfoLocation(InternalNode node)
+        {
+            while (true) {
+                try {
+                    // work for
```
incomplete comment? remove?

phd3 commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -14,9 +14,11 @@
 package io.trino.execution;
```
please shorten the commit message (~72 chars too avoid wrapping) and add more info as commit description: https://cbea.ms/git-commit/

phd3 reviewed (commented):
lgtm, mostly nits

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:
```diff
@@ -70,6 +70,11 @@
     private Duration statusRefreshMaxWait = new Duration(1, TimeUnit.SECONDS);
     private Duration infoUpdateInterval = new Duration(3, TimeUnit.SECONDS);
 
+    private boolean interruptStuckSplitTasksEnabled = true;
```
nit: I wonder if it will be more readable / convenient if the property naming was aligned, e.g.:

- `interruptStuckSplitTasksEnabled` 
- `interruptStuckSplitTasksWarningThreshold`
- `interruptStuckSplitTasksTimeout`
- `interruptStuckSplitTasksDetectionInterval`

WDYT?

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -109,6 +127,7 @@
     private final long queryMaxMemoryPerNode;
 
     private final CounterStat failedTasks = new CounterStat();
+    private final Optional<StuckSplitTaskInterrupter> stuckSplitTaskInterrupter;
```
nit: `stuckSplitTasksInterrupter` (to be consistent with the property naming)

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -597,4 +650,87 @@ public QueryContext getQueryContext(QueryId queryId)
+        return Optional.of(new StuckSplitTaskInterrupter(stuckSplitsWarningThreshold, interruptStuckSplitTasksTimeout, stuckSplitsDetectionInterval, stuckSplitStackTracePredicate, taskExecutor, ticker));
+    }
+
+    private class StuckSplitTaskInterrupter
```
nit: `StuckSplitTasksInterrupter`

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -597,4 +650,87 @@ public QueryContext getQueryContext(QueryId queryId)
+    {
+        private final Duration interruptStuckSplitTasksTimeout;
+        private final Duration stuckSplitsDetectionInterval;
+        private Predicate<List<StackTraceElement>> stuckSplitStackTracePredicate;
```
I would recommend making it final and have a `@VisibleForTesting` version of the `SqlTaskManager` constructor that takes an interrupter that is constructed in a certain way (with a predicate modified for testing)

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -597,4 +650,87 @@ public QueryContext getQueryContext(QueryId queryId)
+            });
+
+            for (TaskId stuckSplitTaskId : stuckSplitTaskIds) {
+                failTask(stuckSplitTaskId, new TrinoException(GENERIC_USER_ERROR, format("Task %s containing long running stuck split get cancelled by the system", stuckSplitTaskId)));
```
Maybe `Task %s is failed`. In Trino cancelled tasks are usually considered to be successful.

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:
```diff
@@ -597,4 +650,87 @@ public QueryContext getQueryContext(QueryId queryId)
+        private void failStuckSplitTasks()
+        {
+            long currentTime = ticker.read();
+            Set<TaskId> stuckSplitTaskIds = this.taskExecutor.getStuckSplitTaskIds((RunningSplitInfo splitInfo) -> {
```
nit: `this.` is not needed (we prefer not to specify `this` if it is not needed)

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java:
```diff
@@ -827,26 +829,33 @@ public long getRunAwaySplitCount()
         private final String threadId;
         private final Thread thread;
         private boolean printed;
+        private final PrioritizedSplitRunner split;
```
nit: Store the task id and split info directly.

arhimondr replied:
Please respond (let me know if you missed it or not going to address)

arhimondr replied:
Discussed offline. The split info cannot be cached as it changes as the split execution progresses.

arhimondr reviewed (commented):
Generally looks good to me % nits.

Please update the commit message according to the guideline: https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#format-git-commit-messages

leetcode-1533 commented:
Hi, I have addressed all the comments. 

I also changed the minimum value for the configs. Due to the async nature, the check based on the walltime is inaccurate when checking splits that runs exceeding 1 second. It can only detect splits that significantly run longer than expected.

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/TaskManagerConfig.java:
```diff
@@ -70,6 +70,11 @@
     private Duration infoUpdateInterval = new Duration(3, TimeUnit.SECONDS);
 
+    private boolean interruptStuckSplitTasksEnabled = true;
+    private Duration interruptStuckSplitTasksWarningThreshold = new Duration(10, TimeUnit.MINUTES);
```
Do we want to set warning threshold to a value lover than the timeout?

arhimondr reviewed (approved):
LGTM % comments

phd3 commented:
Merged, thanks!
